WreckEngine/Text.cpp

```diff
@@ -1,198 +1,571 @@
+        rect.h = static_cast<stbrp_coord>(bitmapHeight + rectPadding);
+        packRects.push_back(rect);
+
+        // Create glyph data
+        auto glyphData = std::make_shared<GlyphData>();
+        auto& glyphMetrics = fontFace->glyph->metrics;
+        glyphData->cp = cp;
+        glyphData->bitmap.resize(rect.w * rect.h, 0); // texture is only alpha values
+        glyphData->glyph.advance = static_cast<float>(fontFace->glyph->metrics.horiAdvance) * kerningScale;
+        glyphData->glyph.bearing = {
+            glyphMetrics.horiBearingX * kerningScale,
+            glyphMetrics.horiBearingY * kerningScale
+        };
+        glyphData->glyph.size = {
+            glyphMetrics.width        * kerningScale,
+            glyphMetrics.height       * kerningScale
+        };
+
+        // Copy glyph buffer data (all pixels are white with varying alpha levels)
+        const unsigned char* pixels = bitmap.buffer;
+        auto& glyphBuffer = glyphData->bitmap;
+        for (uint32_t y = 0; y < bitmapHeight; ++y)
+        {
+            for (uint32_t x = 0; x < bitmapWidth; ++x)
+            {
+                const size_t index = (x + rectPaddingPerSide) + (y + rectPaddingPerSide) * rect.w;
+                if (bitmap.pixel_mode == FT_PIXEL_MODE_MONO)
+                {
+                    // Pixels are 1 bit monochrome values
+                    auto isBitSet = ((pixels[x / 8]) & (1 << (7 - (x % 8))));
+                    glyphBuffer[index] = isBitSet ? 255 : 0;
+                }
+                else
+                {
+                    // Pixels are 8-bit gray scale
+                    glyphBuffer[index] = pixels[x];
+                }
+            }
+            pixels += bitmap.pitch;
+        }
+
+        // Register the glyph data
+        loadedGlyphs.insert({cp, glyphData});
+    }
+
+    // Now let's pack the glyphs into a texture
+    // TODO - Need to ensure packing succeeded. If it didn't, try with a larger texture?
+    std::array<stbrp_node, TEX_SIZE> packNodes;
+    stbrp_context packCtx;
+    stbrp_init_target(&packCtx, TEX_SIZE, TEX_SIZE, packNodes.data(), packNodes.size());
+    stbrp_pack_rects(&packCtx, &packRects[0], packRects.size());
+    
+    // Create the texture
     GL_CHECK(glPixelStorei(GL_UNPACK_ALIGNMENT, 1));
-	for (unsigned char c = 0; c < 128; ++c) {
-		if (FT_Load_Char(fontFace, c, FT_LOAD_RENDER)) { std::cout << "Could not load glyph: " << c << std::endl; continue; }
-		auto& glyph = fontFace->glyph;
-		auto& bitmap = glyph->bitmap;
-		GLtexture t;
-		t.create();
-		t.bind();
-		t.set2D<GLubyte>(bitmap.buffer, bitmap.width, bitmap.rows, GL_RED, GL_RED);
-		t.param(GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE); t.param(GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
-		t.param(GL_TEXTURE_MIN_FILTER, GL_LINEAR); t.param(GL_TEXTURE_MAG_FILTER, GL_LINEAR);
-		glyphs[c] = { t, vec2(bitmap.width, bitmap.rows), vec2(glyph->bitmap_left, glyph->bitmap_top), (uint32_t) glyph->advance.x };
-	}
+    tex.bind();
```

Comment:
I was originally under the assumption that calling `create` would re-create the texture and orphan the previous one. I don't really see a problem with having it in the constructor over having it here, though